saged/datasets.py

```diff
@@ -0,0 +1,209 @@
+""" This file contains dataset objects for manipulating gene expression data """
+from abc import ABC, abstractmethod
+from typing import Sequence, Tuple, List
+
+import numpy as np
+
+
+class ExpressionDataset(ABC):
+    """
+    The base dataset defining the API for datasets in this project. This class is
+    based on the Dataset object from pytorch
+    (https://pytorch.org/docs/stable/_modules/torch/utils/data/dataset.html#Dataset
+    """
+    @abstractmethod
+    def __init__(self) -> None:
+        """
+        Abstract initializer. When implemented this function should take in a datasource
+        and use it to initialize the class member variables
+        """
+        pass
+
+    @abstractmethod
+    def __len__(self) -> int:
+        """
+        Return the length of the dataset for use in determining
+        the size of an epoch
+
+        Arguments
+        ---------
+
+        Returns
+        -------
+        length: The number of samples currently available in the dataset
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def __getitem__(self, idx: int) -> Tuple[np.array, np.array]:
+        """
+        Allows access into the dataset to select a single datapoint and its associated label
+
+        Arguments
+        ---------
+        idx: The index of the given item
+
+        Returns
+        -------
+        X: The gene expression data for the given index in a genes x 1 array
+        y: The label corresponding to the sample in X
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_all_data(self) -> Tuple[np.array, np.array]:
+        """
+        Returns all the expression data and labels from the dataset in the
+        form of an (X,y) tuple where both X and y are numpy arrays
+
+        Returns
+        -------
+        X: The gene expression data in a genes x samples array
+        y: The label corresponding to each sample in X
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def subset_samples(self, fraction: float) -> None:
+        """
+        Limit the amount of data available to be returned in proportion to
+        the total amount of data that was present in the dataset.
+        That is to say that calling subset_data_by_fraction twice will not cause the Dataset
+        object to have multiplicatively less data, it will have the fraction of data specified
+        in the second call
+
+        Arguments
+        ---------
+        fraction: The fraction of the samples to keep
```

Comment:
Are you assuming here that the function will modify the data structure that's passed in, rather than creating a copy and modifying and returning that? Or is that up to the inheriting classes?

If you're making an assumption about this, I'd suggest making it clear in the function documentation (same for the `subset_studies` function).

Reply:
I think it makes sense to leave it to the inheriting classes. The function should probably return the Dataset object afterwards either way to allow chaining calls, so I'll add that